File: 54_spiral_matrix/54_spiral_matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function (matrix) {
  const result = [];

  while (matrix.length > 0) {
    result.push(...matrix.shift());

    if (matrix.length > 0 && matrix[0].length > 0) {
      for (let row of matrix) {
        result.push(row.pop());
      }
    }
    if (matrix.length > 0) {
      let row = matrix.pop();
      result.push(...row.reverse());
    }

    if (matrix.length > 0 && matrix[0].length > 0) {
      for (let i = matrix.length - 1; i >= 0; i--) {
        result.push(matrix[i].shift());
      }
    }
  }
  return result;
};

File: 54_spiral_matrix/54_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result = []  # Create an empty list to store the result

        while matrix:  # Continue until the matrix is empty
            # Step 1: Move from left to right and append the first row to the result
            result += matrix.pop(0)

            # Step 2: Move from top to bottom and append the last column to the result
            if matrix and matrix[0]:
                for row in matrix:
                    result.append(row.pop())

            # Step 3: Move from right to left and append the last row (reversed) to the result
            if matrix:
                result += matrix.pop()[::-1]

            # Step 4: Move from bottom to top and append the first column to the result
            if matrix and matrix[0]:
                for row in matrix[::-1]:
                    result.append(row.pop(0))

        return result